File: src/NerdBank.GitVersioning/CloudBuildServices/BitbucketCloud.cs
// Copyright (c) .NET Foundation and Contributors. All rights reserved.
// Licensed under the MIT license. See LICENSE file in the project root for full license information.

namespace Nerdbank.GitVersioning.CloudBuildServices;

/// <summary>
/// Cloud build handling for Bitbucket Cloud.
/// </summary>
/// <remarks>
/// The Bitbucket-specific properties referenced here are <see href="https://support.atlassian.com/bitbucket-cloud/docs/variables-and-secrets/">documented here</see>.
/// </remarks>
public class BitbucketCloud : ICloudBuild
{
    /// <inheritdoc />
    public bool IsApplicable => !string.IsNullOrWhiteSpace(Environment.GetEnvironmentVariable("BITBUCKET_PIPELINE_UUID")) &&
                                !string.IsNullOrWhiteSpace(Environment.GetEnvironmentVariable("BITBUCKET_STEP_UUID")) &&
                                !string.IsNullOrWhiteSpace(Environment.GetEnvironmentVariable("BITBUCKET_STEP_TRIGGERER_UUID"));

    /// <inheritdoc />
    public bool IsPullRequest => !string.IsNullOrWhiteSpace(Environment.GetEnvironmentVariable("BITBUCKET_PR_ID"));

    /// <inheritdoc />
    public string BuildingBranch => CloudBuild.ShouldStartWith(Environment.GetEnvironmentVariable("BITBUCKET_BRANCH"), "refs/heads/");

    /// <inheritdoc />
    public string BuildingTag => CloudBuild.ShouldStartWith(Environment.GetEnvironmentVariable("BITBUCKET_TAG"), "refs/tags/");

    /// <inheritdoc />
    public string GitCommitId => Environment.GetEnvironmentVariable("BITBUCKET_COMMIT");

    /// <inheritdoc />
    public IReadOnlyDictionary<string, string> SetCloudBuildNumber(string buildNumber, TextWriter stdout, TextWriter stderr)
    {
        return new Dictionary<string, string>();
    }

    /// <inheritdoc />
    public IReadOnlyDictionary<string, string> SetCloudBuildVariable(string name, string value, TextWriter stdout, TextWriter stderr)
    {
        return SetCloudBuildVariable(name, value, stdout, stderr, isOutput: false);
    }

    /// <inheritdoc />
    public IReadOnlyDictionary<string, string> SetCloudBuildVariable(string name, string value, TextWriter stdout, TextWriter stderr, bool isOutput)
    {
        // Bitbucket Cloud doesn't support output variables like Azure DevOps, so we ignore the isOutput parameter
        return new Dictionary<string, string>();
    }
}

File: src/nerdbank-gitversioning-rs/src/cloud_build_services/bitbucket_cloud.rs
// Copyright (c) .NET Foundation and Contributors. All rights reserved.
// Licensed under the MIT license. See LICENSE file in the project root for full license information.

//! [Bitbucket Cloud](https://support.atlassian.com/bitbucket-cloud/docs/variables-and-secrets/) support.

use std::io::{self, Write};

use crate::cloud::{CloudBuild, EnvironmentVariables, empty_variables, env, should_start_with};

/// Bitbucket Cloud Pipelines integration.
pub struct BitbucketCloud;

impl CloudBuild for BitbucketCloud {
    fn name(&self) -> &'static str {
        "BitbucketCloud"
    }

    fn is_applicable(&self) -> bool {
        [
            "BITBUCKET_PIPELINE_UUID",
            "BITBUCKET_STEP_UUID",
            "BITBUCKET_STEP_TRIGGERER_UUID",
        ]
        .into_iter()
        .all(|name| env(name).is_some_and(|value| !value.trim().is_empty()))
    }

    fn is_pull_request(&self) -> bool {
        env("BITBUCKET_PR_ID").is_some_and(|value| !value.trim().is_empty())
    }

    fn building_branch(&self) -> Option<String> {
        should_start_with(env("BITBUCKET_BRANCH"), "refs/heads/")
    }

    fn building_tag(&self) -> Option<String> {
        should_start_with(env("BITBUCKET_TAG"), "refs/tags/")
    }

    fn git_commit_id(&self) -> Option<String> {
        env("BITBUCKET_COMMIT")
    }

    fn set_cloud_build_number(
        &self,
        _build_number: &str,
        _stdout: &mut dyn Write,
        _stderr: &mut dyn Write,
    ) -> io::Result<EnvironmentVariables> {
        Ok(empty_variables())
    }

    fn set_cloud_build_variable(
        &self,
        _name: &str,
        _value: &str,
        _stdout: &mut dyn Write,
        _stderr: &mut dyn Write,
    ) -> io::Result<EnvironmentVariables> {
        Ok(empty_variables())
    }
}
